modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -137,6 +150,153 @@ public IgniteRel visit(IgniteTableScan rel) {
         return rel;
     }
 
+    private static class ModifyNodeShuttle extends IgniteRelShuttle {
+        List<RexNode> projections = null;
+        IgniteValues values = null;
+        List<List<RexNode>> exprProjections = null;
+        boolean collectExprProjections = false;
+
+        /** {@inheritDoc} */
+        @Override
+        public IgniteRel visit(IgniteProject rel) {
+            if (projections == null && !collectExprProjections) {
+                projections = rel.getProjects();
+            } else {
+                if (exprProjections == null) {
+                    exprProjections = new ArrayList<>();
+                }
+                exprProjections.add(rel.getProjects());
+            }
+            return super.visit(rel);
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public IgniteRel visit(IgniteValues rel) {
+            if (!collectExprProjections) {
+                values = rel;
+            }
+            return super.visit(rel);
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public IgniteRel visit(IgniteUnionAll rel) {
+            collectExprProjections = true;
+            return super.visit(rel);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public IgniteRel visit(IgniteTableModify rel) {
+        if (rel.getOperation() != INSERT && rel.getOperation() != DELETE) {
```

Comment:
why DELETE is different from UPDATE? Btw, what about MERGE? Do we have pruning tests for MERGE?

Reply:
UPDATE statement is make sense  if update predicate contains all keys from pk and is such a case it can be processed by other branch :
```
      IgniteTableModify(table=[[PUBLIC, T]], operation=[UPDATE], updateColumnList=[[C2]]
        IgniteTableScan(table=[[PUBLIC, T]], filters=[=($t0, 42)]
```
Pruning for MERGE is not implemented at all, what kind of MERGE request it would be helpful ?

Reply:
done